typescript/radix-sdk/src/modules/query.ts

```diff
@@ -0,0 +1,769 @@
+import { BigNumber } from 'bignumber.js';
+import { utils } from 'ethers';
+
+import { assert, ensure0x } from '@hyperlane-xyz/utils';
+
+export class RadixQuery {
+  protected networkId: number;
+  protected gateway: GatewayApiClient;
+
+  constructor(networkId: number, gateway: GatewayApiClient) {
+    this.networkId = networkId;
+    this.gateway = gateway;
+  }
+
+  public async getXrdAddress() {
+    const knownAddresses = await LTSRadixEngineToolkit.Derive.knownAddresses(
+      this.networkId,
+    );
+    return knownAddresses.resources.xrdResource;
+  }
+
+  public async isGatewayHealthy(): Promise<boolean> {
+    const status = await this.gateway.status.getCurrent();
+    return status.ledger_state.state_version > 0;
+  }
+
+  public async estimateTransactionFee({
+    transactionManifest,
+  }: {
+    transactionManifest: TransactionManifest;
+  }): Promise<{ gasUnits: bigint; gasPrice: number; fee: bigint }> {
+    const pk = new PrivateKey.Ed25519(new Uint8Array(utils.randomBytes(32)));
+    const constructionMetadata =
+      await this.gateway.transaction.innerClient.transactionConstruction();
+
+    const manifest = (
+      await RadixEngineToolkit.Instructions.convert(
+        transactionManifest.instructions,
+        this.networkId,
+        'String',
+      )
+    ).value as string;
+
+    const response =
+      await this.gateway.transaction.innerClient.transactionPreview({
+        transactionPreviewRequest: {
+          manifest,
+          nonce: generateRandomNonce(),
+          signer_public_keys: [
+            {
+              key_type: 'EddsaEd25519',
+              key_hex: pk.publicKeyHex(),
+            },
+          ],
+          flags: {
+            use_free_credit: true,
+            // we have to enable this flag because the signer of the tx is a random pk
+            // this allows us to simulate txs for different addresses - even if we don't have accesse to their public key
+            assume_all_signature_proofs: true,
+          },
+          start_epoch_inclusive: constructionMetadata.ledger_state.epoch,
+          end_epoch_exclusive: constructionMetadata.ledger_state.epoch + 2,
+        },
+      });
+
+    assert(
+      !(response.receipt as any).error_message,
+      `${(response.receipt as any).error_message}`,
+    );
+
+    const fee_summary: FeeSummary = (response.receipt as any).fee_summary;
+    const costing_parameters: CostingParameters = (response.receipt as any)
+      .costing_parameters;
+
+    const gasUnits =
+      BigInt(fee_summary.execution_cost_units_consumed) +
+      BigInt(fee_summary.finalization_cost_units_consumed);
+    const fee = BigInt(
+      new BigNumber(fee_summary.xrd_total_execution_cost)
+        .plus(BigNumber(fee_summary.xrd_total_finalization_cost))
+        .plus(BigNumber(fee_summary.xrd_total_storage_cost))
+        .times(new BigNumber(10).exponentiatedBy(18))
+        .toFixed(0),
+    );
+    const gasPrice =
```

Comment:
given that gasUnits and fee are both bigint, should this also be a bigint? i can imagine a caller want to do some arithmatic with these values and having to convert